### PR TITLE
Warmboot Manager  HLD for warm reboot improvement 

### PR DESCRIPTION
**Pull request by akarshgupta25**

### CODE REVIEW

**Comment by zhangyanzhao** [comment]
community review recording https://zoom.us/rec/share/G-RZYGwrB0vJXHoIE5pEJ_YsHDdCC4NcbtcEyTG9m7HIPLVpFfO7UOyKeyyfHD6N.72_XWM7l_O90rlno

**Comment by zhangyanzhao** [comment]
202311 release is only for HLD review, no code PR

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (vaibhavhd) [inline doc/warm-reboot/NSF_Manager_HLD.md:164]: Doing this simultaneously to all registered components may not be a good idea. We do want some components to keep operating:
1) until a dependency is active (eg., syncd should work as long as teamd is working)
2) as long as possible (eg., we do not want to freeze teamd even if OA is frozen or swss is stopped).

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (vaibhavhd) [inline doc/warm-reboot/NSF_Manager_HLD.md:174]: IMO, this is not a good idea. Preventing teamd to stop exchanging LACPDUs in phase 1 would put a lot of pressure to reconcile within LAG session standard timeout of 90s. If we spend too much time in phase 2+ then LAGs are guaranteed to flap in the recovery path.

Comment (vaibhavhd) [reply doc/warm-reboot/NSF_Manager_HLD.md:174]: Same for syncd - since we want teamd to still continue to send LACPDUs as long as possible, we should not request syncd to reach quiescence at the same time when Orchagent does.

Comment (vaibhavhd) [reply doc/warm-reboot/NSF_Manager_HLD.md:174]: Thus, I don't think we can have all components reaching `quiescence` as a prerequisite for going into Phase 2.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (vaibhavhd) [inline doc/warm-reboot/NSF_Manager_HLD.md:184]: > NSF Manager will wait for a period of time for the switch to become quiescent after which it will determine that warm reboot failed and abort the warm reboot operation.

The design should also account for restoration of components state (unfreeze) in the event of failure.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (vaibhavhd) [inline doc/warm-reboot/NSF_Manager_HLD.md:193]: How is this phase different from what was done as last step in Phase 1? Also, please define `consistent state`.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (vaibhavhd) [inline doc/warm-reboot/NSF_Manager_HLD.md:220]: > The design allows applications to transition through these phases independently

But in warm shutdown we do need dependent transition. 
`independent` transition can cause confusion and open gates for multiple race conditions.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (qiluo-msft) [inline doc/warm-reboot/NSF_Manager_HLD.md:1]: Could you prevent using the terminalogy "NSF" in this design doc? There is another active HLD relating gNOI API, and we are discussing use gNOI NSFT terminalogy to represent SONIC fast-reboot instead of SONiC warm-reboot, since there is another gNOI WARM already. ref: https://github.com/openconfig/gnoi/blob/98d6b81c6dfe3c5c400209f82a228cf3393ac923/system/system.proto#L128

May I propose "Warm Reboot Manager" in this HLD?

**Review by qiluo-msft — DISMISSED** [review]
As comment.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:99]: This is referring to the separate binaries that we have to send shutdown notifications to different components. For example, _orchagent_restart_check_ for Orchagent freeze notification, _syncd_request_shutdown_ for Syncd notifications etc.

We aren't proposing to have separate service management scripts.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:100]: swss-common enables us to create a pub/sub model wherein we can send notifications and subscribe for warm-boot state updates from multiple applications in Redis DB as opposed to poll for this information for individual components.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:101]: The shutdown state machine for different components (proposed in this HLD) can be better monitored using a daemon that is leveraging swss-common libs as opposed to polling these warm-boot states of individual components using a bash script.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:105]: This framework will co-exist with the existing shutdown orchestration. This framework introduces additional notification channels for the newer unified shutdown notifications. So components can choose to take actions based on the notification channel from which they receive the shutdown notification. Our plan is to be compatible with the existing SONiC warm-boot orchestration.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:126]: Can you please elaborate on the type of fixes that are introduced at runtime?

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:164]: Freeze notification means that components should stop generating events for which they are the source. All components will continue to serve requests received by them but will stop generating events that are originated by them. This is similar to saying "I will not ask questions, but I will continue to answer questions". This means that components will continue to serve internal requests. After a point of time, all components will stop generating events and thus the switch will become quiescent.

This means that during freeze phase, Orchagent and Syncd will continue to serve requests from other components such as Teamd.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:174]: I understand the concern here. There are multiple ways to tackle this problem:
1. After freeze notification, Teamd continue to exchange LACP PDUs but not forward any request to Orchagent/Syncd so that their quiescence doesn't change.
2. Setup a control plane assistant that serves as a proxy for Teamd.
3. Increase the LAG session timeout before stopping LACP PDU exchange.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:184]: Does SONiC currently support unfreeze of applications in case of failures?

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:193]: Phase 1 ensures that the switch reaches a stable state. Phase 2 (optional) audits the switch state. State verification enables auditing of the state of different components and ensuring that they are in an expected state. Consistency depends on the component. It can mean that the internal cache matches the switch state, or the switch intent matches the switch state etc. For example, BGP kernel routes match APPL DB information, oper-status in Linux host interface matches APPL DB information etc.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:220]: Phase 1 ensures that the switch stack components continue to serve requests from other components. Also, having a state machine during shutdown helps alleviate the race conditions. I would be happy to look at individual examples of race conditions and check if this design is able to handle them.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (srideepDell) [inline doc/warm-reboot/NSF_Manager_HLD.md:256]: Could add note that feature/new changes are independent of any SAI changes needed. There could be possible SAI changes needed for improvement.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (srideepDell) [inline doc/warm-reboot/NSF_Manager_HLD.md:29]: Couple of sections can be added. one use cases covered ( LACP.. etc). Also sections which NPU types are tested or supported after testing.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:12]: Added details in the Architecture Design section.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:128]: Updated freeze and reconciliation images. Hope they are readable now.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:149]: These couple of lines provide the brief definition of these flags. Do more details need to be added?

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:29]: Added NPU types detail in Testing Requirements/Design section.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:264]: Added a note in SAI API section.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:170]: No, boundary applications that are independent of state changes to other switch components such as UMF, P4RT etc. can aggregate Phase 1, 2 and 3 as a part of the freeze notification (see Application Shutdown Optimization section). But this is applicable for only a certain subset of the switch stack components. NSF Manager needs to wait for the entire switch to become quiescent before a global checkpointing can take place. For example, Syncd should only save internal states after all intent has been programmed i.e. the switch is quiescent.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:204]: This is an interesting point :)

Some context:
For some switch stack components such as Syncd, checkpoint notification results in saving the internal state, disconnecting from the ASIC and exiting. This is similar to processing the `shutdown` notification being sent by `/usr/bin/syncd_request_shutdown` binary in the existing warm-boot orchestration.
Additionally, we are exploring whether we should support unfreeze during warm shutdown. One possible situation is that we send freeze, then state verification and unfreeze on state verification failure.

With this background, if we support state verification and checkpoint as a single phase then Syncd can exit after checkpointing and state verification can fail for some other application. In such a case, we wouldn't be able to freeze --> state verification failure --> unfreeze unless we ensure that checkpointing isn't a point of no return for a switch stack component.

So we need to analyze the pros and cons of merging these 2 phases. Let me discuss it internally and get back.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:256]: Since this is an initial NSF Manager design, the current proposal is to abort warm reboot if an application reports _failed_ warm-boot state. However, we are exploring `unfreeze` on warm-boot failure as well. For example, if Phase 1 (freeze) fails then we can unfreeze the switch stack and return that warm reboot failed.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md:175]: This design introduces the high-level principles for the new warm-boot orchestration framework. I wanted to limit the scope of this design to these core principles. That is why this section talks about the actions of different components at a very high-level, thereby providing a brief idea of stopping self-sourced events.

From P4RT perspective, the P4Runtime spec doesn't talk about reboots in general. During cold reboot, the client observes the same behavior as a server shutdown. Ideally, the client shouldn't observe different behavior during warm reboot. This is because from the P4Runtime lens, the server was shutdown (thus unavailable for sometime) and came back up in the same state as before thereby maintaining the read-write symmetry. I would be happy to discuss this further in the System Orchestration workgroup or in a separate thread.

**Review by akarshgupta25 — COMMENTED** [review]
Addressed Ayyappa's comments.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (qiluo-msft) [reply doc/warm-reboot/NSF_Manager_HLD.md:1]: Could you prevent using the terminology "NSF" in this design doc?

**Comment by zhangyanzhao** [comment]
@akarshgupta25 can you please help to add the code PR to this HLD by referring to #806 ? Thanks.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (qiluo-msft) [reply doc/warm-reboot/NSF_Manager_HLD.md:1]: I somehow prefer using gnoi WARM terminology over NSF. Would you mind change all the NSF->WARM in this HLD including PR title and markdown headers. I will not block this PR if using WARM in HLD.

**Comment by zhangyanzhao** [comment]
code PR is not ready, move to backlog for future release

**Review by vaibhavhd — APPROVED** [review]
We need a test plan HLD too. You had mentioned that you are working on it, do you have it ready for review now?
I wouldn't say that the testplan is out of scope for this design. SONiC works on TDD. So I would like to have test plan ready before we start merging code changes.

Just making a note: this feature is backward compatible by design. But that also needs to be reflected by the design changes needed in critical containers - swss, bgp, syncd, teamd etc. If the change in these services is not backward compatible, then the NSF manager design is not backward compatible as a whole.

File: doc/warm-reboot/NSF_Manager_HLD.md
Comment (akarshgupta25) [reply doc/warm-reboot/NSF_Manager_HLD.md]: Replaced NSF with Warmboot. Updated the document tile, headings and diagrams.

Comment (qiluo-msft) [reply doc/warm-reboot/NSF_Manager_HLD.md]: Thanks!

**Review by anuthalmrvl — APPROVED** [review]
Thanks @akarshgupta25 for multiple detailed review sessions and addressing the comments.

**Review by vaibhavhd — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Review by srideepDell — APPROVED** [review]

**Comment by bhagatgit** [comment]
Received approvals from Microsoft, Dell and Marvell

**Comment by liat-grozovik** [comment]
@bhagatgit is there a plan to have the code PRs in 202411? 
All vendors need to be ready to test it before it is merged and ensure no regression with this approach.
I dont see any reference to the app extension. any reason why the new approach is not taking it at least in the design part?

**Comment by rkavitha-hcl** [comment]
| Repo    | PR title |State|
| -------- | ------- |------- |
| sonic-buildimage  | [gNOI Cold Reboot - Addition of new framework docker ](https://github.com/sonic-net/sonic-buildimage/pull/20709)   |Open |
| sonic-buildimage|   [gNOI Cold Reboot - Integrated Rebootbackend changes](https://github.com/sonic-net/sonic-buildimage/pull/20710)  |Open|
| sonic-buildimage    | [gNOI Cold Reboot - Integrated tests](https://github.com/sonic-net/sonic-buildimage/pull/20711) |Open|
| sonic-buildimage    |[gNOI Cold Reboot - Integrated gNOI repository](https://github.com/sonic-net/sonic-buildimage/pull/20712)|Open|
| sonic-buildimage    |[gNOI Cold Reboot - Integrated Docker and Makefile changes](https://github.com/sonic-net/sonic-buildimage/pull/20713)|Open|
| sonic-buildimage    |[gNOI Warm Reboot - rebootbackend changes](https://github.com/sonic-net/sonic-buildimage/pull/20786)|Open|
| sonic-buildimage    |[gNOI Warm Reboot - Added tests](https://github.com/sonic-net/sonic-buildimage/pull/20801)|Open|
| sonic-host-services    |[sonic-host-services changes for gNOI Cold Reboot](https://github.com/sonic-net/sonic-host-services/pull/181)|Open|
| sonic-host-services    |[sonic-host-services changes for gNOI Warm Reboot](https://github.com/sonic-net/sonic-host-services/pull/191)|Open|

**Comment by bibhuprasad-hcl** [comment]
[Addition of new framework docker](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20709)

[Integrated Rebootbackend code changes to handle Cold Reboot feature.](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20710)

[Added UT cases for Cold Reboot feature.](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20711)

[Added gNOI repository as a  submodule and modified the import paths for the proto files.](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20712)

[Integrated Docker and Makefile changes.](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20713)

[Enabled Rebootbackend to support gNOI Warm Reboot.](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20786)

[Added UT cases for warm reboot feature.](https://img.shields.io/github/pulls/detail/state/Azure/sonic-buildimage/20801)

[Added sonic-host-services changes for gNOI Cold Reboot](https://img.shields.io/github/pulls/detail/state/Azure/sonic-host-services/181)

[Added sonic-host-services changes for gNOI Cold/warm Reboot](https://img.shields.io/github/pulls/detail/state/Azure/sonic-host-services/186)

[Added incemental code change for sonic-host-services changes for gNOI Warm Reboot](https://img.shields.io/github/pulls/detail/state/Azure/sonic-host-services/191)

**Pull request merged**
